cpp/monitoring/gcm/exporter.cc

```diff
@@ -0,0 +1,264 @@
+  token_refreshed_at_ = system_clock::now();
+
+  JsonObject reply(resp.body);
+  CHECK(reply.Ok()) << "Failed to parse metadata JSON:\n" << resp.body;
+  JsonString bearer(reply, "access_token");
+  CHECK(bearer.Ok());
+  bearer_token_ = bearer.Value();
+
+  VLOG(1) << "GCM credentials refreshed";
+}
+
+
+namespace {
+
+
+void AddLabelDescription(const string& key, const string& desc,
+                         JsonArray* labels) {
+  JsonObject l;
+  l.Add("key", kCloudPrefix + key);
+  l.Add("description", desc);
+  CHECK_NOTNULL(labels)->Add(&l);
+}
+
+
+}  // namespace
+
+
+void GCMExporter::CreateMetrics() const {
+  const std::set<const Metric*> metrics(Registry::Instance()->GetMetrics());
+  for (auto& m : metrics) {
+    CHECK_NOTNULL(m);
+
+    // See
+    // https://cloud.google.com/monitoring/v2beta2/metricDescriptors#resource
+    // for a description of the structure we're building here.
+
+    JsonArray labels;
+    AddLabelDescription("instance",
+                        "Instance from which the sample originates.", &labels);
+    for (const auto& label : m->LabelNames()) {
+      AddLabelDescription(label, label, &labels);
+    }
+
+    JsonObject desc;
+    switch (m->Type()) {
+      case Metric::COUNTER:
+        // only gauge type metrics are supported for custom metrics currently:
+        // https://cloud.google.com/monitoring/api/metrics#metric-types
+        desc.Add("metricType", "gauge");
+        break;
+      case Metric::GAUGE:
+        desc.Add("metricType", "gauge");
+        break;
+      default:
+        LOG(FATAL) << "Unknown type: " << m->Type();
+    }
+    desc.Add("valueType", "double");
+
+    JsonObject metric;
+    metric.Add("name", kCloudPrefix + m->Name());
+    metric.Add("description", m->Help());
+    metric.Add("labels", labels);
+    metric.Add("typeDescriptor", desc);
+
+    UrlFetcher::Request req(
+        (URL(FLAGS_google_compute_monitoring_metrics_url)));
+    req.verb = UrlFetcher::Verb::POST;
+    req.headers.insert(make_pair("Content-Type", "application/json"));
+    req.headers.insert(make_pair("Authorization", "Bearer " + bearer_token_));
+    req.body = metric.ToString();
+
+    UrlFetcher::Response resp;
+    SyncTask task(executor_);
+    VLOG(1) << "Creating metric m.Name()...";
+    VLOG(2) << req.body;
+    fetcher_->Fetch(req, &resp, task.task());
+    task.Wait();
+    CHECK_EQ(util::Status::OK, task.status());
+    CHECK_EQ(200, resp.status_code);
+    VLOG(1) << "Metrics Created.";
+    VLOG(2) << resp.body;
+  }
+}
+
+
+namespace {
+
+
+void AddLabel(const string& key, const string& value, JsonObject* labels) {
+  CHECK_NOTNULL(labels)->Add((kCloudPrefix + key).c_str(), value);
+}
+
+
+}  // namespace
+
+
+void GCMExporter::PushMetrics(util::Task* child_task) {
+  if (task_.task()->CancelRequested()) {
+    task_.task()->Return(util::Status::CANCELLED);
+    return;
+  }
+
+  if (system_clock::now() - token_refreshed_at_ > minutes(3)) {
+    RefreshCredentials();
```

Comment:
I've added a flag and dropped the default to 30m (I think they expire after 60 by default.)